### PR TITLE
eth/catalyst: fix races

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**

### CODE REVIEW

File: eth/catalyst/api_test.go [hunk 1]
```diff
@@ -1240,9 +1240,10 @@ func TestNilWithdrawals(t *testing.T) {
 
 func setupBodies(t *testing.T) (*node.Node, *eth.Ethereum, []*types.Block) {
 	genesis, blocks := generateMergeChain(10, true)
-	n, ethservice := startEthService(t, genesis, blocks)
 	// enable shanghai on the last block
-	ethservice.BlockChain().Config().ShanghaiTime = &blocks[len(blocks)-1].Header().Time
+	time := blocks[len(blocks)-1].Header().Time + 1
```
Comment (holiman) [inline eth/catalyst/api_test.go:1244]: This does something more than just fix a races, it also bumps the time by `1` second. So it also either fixes or introduces a bug... ?

Comment (MariusVanDerWijden) [reply eth/catalyst/api_test.go:1244]: It fixes a bug. Previously shanghai was enabled `before` the last block, now its enabled afterwards. This means previously we would import the block on pre-shanghai rules, then afterwards set the chain config so the last block is post-shanghai. Basically we imported the block and modified the rules afterwards.

**Review by holiman — APPROVED** [review]
LGTM

**Comment by MariusVanDerWijden** [comment]
@holiman I added another race fix on top, now the race test does not show any races in this package anymore. The fix for the second race was very similar to the first one. In order to check that we correctly return INVALID_TTD we need to create a post-merge block with an invalid ttd.
The old version did this by setting the ttd lower and creating a new block. The new code just requests a block on top of an older block. However the blockhash does not match, because the miner computes the blockhash correctly, thus we need to recompute the blockhash before sending the invalid payload

File: eth/catalyst/api_test.go [hunk 2]
```diff
@@ -913,6 +908,28 @@ func TestNewPayloadOnInvalidTerminalBlock(t *testing.T) {
...
+		Number:      new(big.Int).SetUint64(data.Number),
+		GasLimit:    data.GasLimit,
+		GasUsed:     data.GasUsed,
+		Time:        data.Timestamp,
+		BaseFee:     data.BaseFeePerGas,
+		Extra:       data.ExtraData,
+		MixDigest:   data.Random,
+	}
+	block := types.NewBlockWithHeader(header).WithBody(txs, nil /* uncles */)
+	data.BlockHash = block.Hash()
```
[the remark this answers is not in the document]
Comment (holiman) [reply eth/catalyst/api_test.go:931]: Why did you mark this as resolved?

**Pull request merged**